platforms/core-execution/build-cache-local/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCache.java

```diff
@@ -43,17 +44,21 @@
 @NonNullApi
 public class DirectoryBuildCache implements BuildCacheTempFileStore, Closeable, LocalBuildCache {
 
-    private final PathKeyFileStore fileStore;
     private final PersistentCache persistentCache;
     private final BuildCacheTempFileStore tempFileStore;
     private final FileAccessTracker fileAccessTracker;
     private final String failedFileSuffix;
     private final ReadWriteLock lock = new ReentrantReadWriteLock();
 
-    public DirectoryBuildCache(PathKeyFileStore fileStore, PersistentCache persistentCache, BuildCacheTempFileStore tempFileStore, FileAccessTracker fileAccessTracker, String failedFileSuffix) {
-        this.fileStore = fileStore;
+    public DirectoryBuildCache(PersistentCache persistentCache, FileAccessTracker fileAccessTracker, String failedFileSuffix) {
         this.persistentCache = persistentCache;
-        this.tempFileStore = tempFileStore;
+        this.tempFileStore = new DefaultBuildCacheTempFileStore((prefix, suffix) -> {
+            try {
+                return Files.createTempFile(persistentCache.getBaseDir().toPath(), prefix, suffix).toFile();
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
```

Comment:
💭 The directory build cache creates the cache artifact files (the TAR.GZs) in a temporary location, and uses atomic file moves to move them into the cache directory once they are finished. We used to create them in a separate temporary directory, which happened to be in the Gradle user home.

However, this can lead to problems, when the local build cache and the Gradle user home were on different file systems. In that case we silently did not move anything to the cache directory. Now we fail in such a case, which led to uncovering this problem.

We now have switched to creating the temporary cache artifact files inside the cache directory, ensuring that they are always on the same file system.

Reply:
Good point, added a comment